harper-core/src/parsers/cases.rs

```diff
@@ -0,0 +1,247 @@
+use crate::Lrc;
+use std::collections::VecDeque;
+
+use itertools::Itertools;
+
+use super::{Parser, TokenKind};
+use crate::patterns::{PatternExt, SequencePattern};
+use crate::{Dictionary, FullDictionary, MergedDictionary, Span, Token, VecExt, WordMetadata};
+
+/// A parser that wraps any other parser to collapse token strings that match
+/// the pattern word_word or word-word.
+pub struct Cases {
+    inner: Box<dyn Parser>,
+    dict: Lrc<MergedDictionary<FullDictionary>>,
+}
+
+impl Cases {
+    pub fn new(inner: Box<dyn Parser>, dict: &Lrc<MergedDictionary<FullDictionary>>) -> Self {
+        Self {
+            inner,
+            dict: dict.clone(),
+        }
+    }
+}
+
+thread_local! {
+    static WORD_OR_NUMBER: Lrc<SequencePattern> = Lrc::new(SequencePattern::default()
+                .then_any_word()
+                .then_one_or_more(Box::new(SequencePattern::default()
+        .then_case_separator()
+        .then_any_word())));
+}
+
+impl Parser for Cases {
```

Comment:
I'm currently getting a compile time error when building for `harper-cli`:
```rust
error[E0277]: `Rc<MergedDictionary<FullDictionary>>` cannot be shared between threads safely
  --> harper-core/src/parsers/cases.rs:34:17
   |
34 | impl Parser for Cases {
   |                 ^^^^^ `Rc<MergedDictionary<FullDictionary>>` cannot be shared between threads safely
   |
   = help: within `Cases`, the trait `Sync` is not implemented for `Rc<MergedDictionary<FullDictionary>>`, which is required by `Cases: Sync`
note: required because it appears within the type `Cases`
  --> harper-core/src/parsers/cases.rs:12:12
   |
12 | pub struct Cases {
   |            ^^^^^
note: required by a bound in `parsers::Parser`
  --> harper-core/src/parsers/mod.rs:15:26
   |
15 | pub trait Parser: Send + Sync {
   |                          ^^^^ required by this bound in `Parser`

error[E0277]: `Rc<MergedDictionary<FullDictionary>>` cannot be sent between threads safely
  --> harper-core/src/parsers/cases.rs:34:17
   |
34 | impl Parser for Cases {
   |                 ^^^^^ `Rc<MergedDictionary<FullDictionary>>` cannot be sent between threads safely
   |
   = help: within `Cases`, the trait `Send` is not implemented for `Rc<MergedDictionary<FullDictionary>>`, which is required by `Cases: Send`
note: required because it appears within the type `Cases`
  --> harper-core/src/parsers/cases.rs:12:12
   |
12 | pub struct Cases {
   |            ^^^^^
note: required by a bound in `parsers::Parser`
  --> harper-core/src/parsers/mod.rs:15:19
   |
15 | pub trait Parser: Send + Sync {
   |                   ^^^^ required by this bound in `Parser`

For more information about this error, try `rustc --explain E0277`.
```

Reply:
This was an issue I was having with tests when running it in the `harper-core` repository, it has to do with the `Lrc` definition, but I'm not quite sure how to fix it.

Reply:
I implemented a fix... but obviously no warnings doesn't mean I didn't screw something up performance-wise. I'll need to get elijah to look at it to figure out what's going on with the concurrent feature.